File: datastructure&algorithm/ds_algo_java/src/daily/d_2022_10_20/bj_2644/Main.java
package daily.d_2022_10_20.bj_2644;
import java.io.*;
import java.io.OutputStreamWriter;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

import java.util.ArrayList;


public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());//총 인원

        //촌수를 계산할 두 사람 a와 b
        StringTokenizer st = new StringTokenizer(br.readLine());
        int a = Integer.parseInt(st.nextToken());
        int b = Integer.parseInt(st.nextToken());

        //그래프
        ArrayList<Integer>[] fam = new ArrayList[n+1];
        for(int i=1;i<=n;i++){
            fam[i] = new ArrayList<>();
        }

        int nodeNum = Integer.parseInt(br.readLine());//촌수
        for(int i=0;i<nodeNum;i++){
            st = new StringTokenizer(br.readLine());
            int s = Integer.parseInt(st.nextToken());//시작
            int e = Integer.parseInt(st.nextToken());//끝
            fam[s].add(e);
            fam[e].add(s);
        }

        boolean[] visited = new boolean[n+1];
        int result = bfs(fam,visited, a,b);
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        bw.write(Integer.toString(result));
        bw.flush();
        bw.close();
    }

    static int bfs(ArrayList<Integer>[] fam, boolean[] visited, int start, int end){
        Queue<int[]> q = new LinkedList<int[]>();
        q.offer(new int[]{start,0});
        visited[start] = true;

        while(!q.isEmpty()){
            int[] cur = q.poll();
            int nowMan = cur[0];//현재 가족
            int chon = cur[1];//현재 촌수

            ArrayList<Integer> nodes = fam[nowMan];//현재가족의 노드들
            for(int n:nodes){
                if(visited[n]) continue;
                if(n==end) return chon+1;
                visited[n]=true;
                q.offer(new int[]{n,chon+1});

            }
        }
        return -1;


    }
}

File: datastructure&algorithm/ds_algo_java/src/daily/d_2022_10_20/bj_2644/Main.kt
package daily.d_2022_10_20.bj_2644
import java.util.Queue;
import java.util.LinkedList;

private fun readInts() = readLine()!!.split(" ").map{it.toInt()}
fun main(){
    //총 가족 수 만큼 맵 생성
    val num = readLine()!!.toInt()
    val map = Array<MutableList<Int>>(num+1){ mutableListOf() }

    //촌수 구할 두명
    val(start,end) = readInts()

    //총 엣지 수
    val edges = readLine()!!.toInt()
    for(i in 1..edges){
        val(s,e) = readInts()
        map[s].add(e)
        map[e].add(s)
    }

    val visited = Array<Boolean>(num+1){false}

    println(bfs(map,visited,start,end))
}

fun bfs(map:Array<MutableList<Int>>, visited:Array<Boolean>, start:Int, end:Int):Int{
    val q:Queue<Array<Int>> = LinkedList<Array<Int>>()
    q.offer(arrayOf(start,0))
    visited[start]=true

    while(!q.isEmpty()){
        val curNode = q.poll()
        val man = curNode[0]
        val chon = curNode[1]

        for(i in map[man]){
            if(visited[i]) continue
            if(i==end) return chon+1

            visited[i]=true
            q.offer(arrayOf(i,chon+1))
        }
    }
    return -1
}